File: app/src/main/java/com/rogger/bp/ui/home/HomeFragment.java
package com.rogger.bp.ui.home;

import static com.rogger.bp.ui.home.CustomProgressBarKt.hideLoadingDialog;
import static com.rogger.bp.ui.home.CustomProgressBarKt.showLoadingDialog;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;
import android.widget.ViewFlipper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.rogger.bp.R;
import com.rogger.bp.data.model.Categoria;
import com.rogger.bp.data.model.Produto;
import com.rogger.bp.databinding.FragmentHomeBinding;
import com.rogger.bp.notification.NotificationPrefs;
import com.rogger.bp.ui.base.CategoriaDialogUtil;
import com.rogger.bp.ui.base.Utils;
import com.rogger.bp.ui.viewmodel.CategoriaViewModel;
import com.rogger.bp.ui.viewmodel.DataViewModel;

import java.util.List;

public class HomeFragment extends Fragment implements OnItemClickListener {
    private ViewFlipper viewFlipper;
    private ImageView imageView;
    private FragmentHomeBinding binding;
    private DataViewModel dataViewModel;
    private CategoriaViewModel categoriaViewModel;
    private AdapterHome adapte;
    private NavController navController;
    private Bundle bundle;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {


        binding = FragmentHomeBinding.inflate(inflater, container, false);

        RecyclerView recyclerView = binding.rcViewListHome;
        viewFlipper = binding.viewFlipper;
        imageView = binding.imgHomeFragment;

        // Recuperar o valor do slider de dias amarelo do SharedPreferences
        int diasAmarelo = NotificationPrefs.getDays(requireContext());

        // Instanciar o AdapterHome com o valor dinâmico do slider
        adapte = new AdapterHome(requireContext(), diasAmarelo, this);

        recyclerView.setAdapter(adapte);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));

        // Inicializar DataViewModel
        dataViewModel = new ViewModelProvider(this).get(DataViewModel.class);

        // ✅ Observar o estado de carregamento (Firebase Sync)
        dataViewModel.getIsLoading().observe(getViewLifecycleOwner(), isLoading -> {
            if (Boolean.TRUE.equals(isLoading)) {
                showLoadingDialog(requireContext());
            } else {
                hideLoadingDialog();
            }
        });

        // Observar os dados do banco de dados
        dataViewModel.getProdutos().observe(getViewLifecycleOwner(), produtos -> {
            if (produtos != null && !produtos.isEmpty()) {
                adapte.setDados(produtos);
                adapte.ordenarPorDiferencaDeDias();
                atualizarView(produtos);
            } else {
                atualizarView(null);
            }
        });

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        // ✅ ViewModel inicializado uma única vez aqui, não dentro de mostrarDialogoSelecaoCategoria()
        categoriaViewModel = new ViewModelProvider(this).get(CategoriaViewModel.class);

        FloatingActionButton fab = binding.fab;
        fab.setOnClickListener(view1 -> {
            mostrarDialogoSelecaoCategoria();
        });
        super.onViewCreated(view, savedInstanceState);
    }

    /**
     * Exibe um diálogo para o utilizador selecionar uma categoria antes de ir para o scanner.
     * Usa observeOnce para garantir que o diálogo abre apenas uma vez por clique,
     * sem acumular observers a cada chamada.
     */
    private void mostrarDialogoSelecaoCategoria() {
        observeOnce(categoriaViewModel.getCategories(), categorias -> {
            if (categorias == null || categorias.isEmpty()) {
                mostrarDialogoAdicionarCategoria();
            } else {
                mostrarDialogoComSpinner(categorias);
            }
        });
    }

    /**
     * Observa um LiveData apenas uma vez e remove o observer automaticamente.
     */
    private <T> void observeOnce(androidx.lifecycle.LiveData<T> liveData, androidx.lifecycle.Observer<T> observer) {
        liveData.observe(getViewLifecycleOwner(), new androidx.lifecycle.Observer<T>() {
            @Override
            public void onChanged(T t) {
                liveData.removeObserver(this);
                observer.onChanged(t);
            }
        });
    }

    /**
     * Exibe um diálogo com um Spinner listando as categorias disponíveis
     */
    private void mostrarDialogoComSpinner(List<Categoria> categorias) {
        CategoriaDialogUtil.mostrarDialogo(
                requireContext(),
                categorias,
                new CategoriaDialogUtil.CategoriaCallback() {
                    @Override
                    public void onCategoriaSelecionada(int categoriaId) {
                        irParaScanner(categoriaId);
                    }

                    @Override
                    public void onAdicionarCategoria() {
                        mostrarDialogoAdicionarCategoria();
                    }
                }
        );
    }

    /**
     * Exibe um diálogo para o utilizador criar uma nova categoria
     */
    private void mostrarDialogoAdicionarCategoria() {
        Utils.dialogCategory(
                requireContext(),
                "Criar nova categoria",
                nomeCat -> {
                    Categoria novaCategoria = new Categoria();
                    novaCategoria.setNome(nomeCat);
                    categoriaViewModel.insert(novaCategoria);

                    // ✅ observeOnce evita múltiplas navegações se o LiveData emitir mais de uma vez
                    observeOnce(categoriaViewModel.getCategories(), categorias -> {
                        if (categorias != null) {
                            for (Categoria c : categorias) {
                                if (c.getNome().equals(nomeCat)) {
                                    irParaScanner(c.getId());
                                    break;
                                }
                            }
                        }
                    });
                }
        );
    }

    /**
     * Navega para o BarcodeScanFragment passando o ID da categoria
     */
    private void irParaScanner(int categoriaId) {
        Bundle bundle = new Bundle();
        bundle.putInt("categoria_id", categoriaId);
        if (getView() != null) {
            Navigation.findNavController(requireView()).navigate(R.id.nav_barcode_scan_fragment, bundle);
        }
    }

    private void atualizarView(List<Produto> dados) {
        if (dados == null || dados.isEmpty()) {
            viewFlipper.setDisplayedChild(1); // Mostra a imagem de lista vazia
            imageView.setImageResource(R.drawable.lista_vazia);
            // ✅ Não escondemos o progressBar aqui, deixamos o observer de isLoading cuidar disso
        } else {
            viewFlipper.setDisplayedChild(0); // Mostra o RecyclerView
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onItemClick(int position, List<Produto> data) {
        if (data == null || position >= data.size()) {
            Toast.makeText(getContext(), "Erro ao carregar produto", Toast.LENGTH_SHORT).show();
            return;
        }

        Produto produto = data.get(position);

        // Verificações de null
        int id = produto.getId() != 0 ? produto.getId() : 0;
        bundle = new Bundle();
        navController = Navigation.findNavController(getActivity(), R.id.nav_host_fragment_content_main);
        bundle.putInt("id", id);
        navController.navigate(R.id.action_nav_home_to_nav_edit_fragment, bundle);
    }

    @Override
    public void onImageClick(String uri) {
        bundle = new Bundle();
        navController = Navigation.findNavController(getActivity(), R.id.nav_host_fragment_content_main);
        bundle.putString("imageUri", uri);
        navController.navigate(R.id.action_nav_home_to_nav_show_fragment, bundle);
    }
}

File: app/src/main/java/com/rogger/bp/ui/home/view/HomeFragment.kt
package com.rogger.bp.ui.home.view

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.snackbar.Snackbar
import com.rogger.bp.R
import com.rogger.bp.data.database.BpDatabase
import com.rogger.bp.data.model.PostCategory
import com.rogger.bp.data.model.PostProduct
import com.rogger.bp.databinding.FragmentHomeBinding
import com.rogger.bp.notification.NotificationPrefs
import com.rogger.bp.ui.base.CategoriaDialogUtil
import com.rogger.bp.ui.base.DialogUtil
import com.rogger.bp.ui.commun.DependencyInjector
import com.rogger.bp.ui.commun.SharedPreferencesManager
import com.rogger.bp.ui.home.ContractHome
import com.rogger.bp.ui.home.CustomProgressBar
import com.rogger.bp.ui.home.OnItemClickListener
import com.rogger.bp.ui.home.presentation.HomePresenter
import kotlinx.coroutines.launch

/*
 * Desenvolvido por Roger de Oliveira
 * Data: 13/05/2026
 * Hora: 22:36
 */
class HomeFragment : Fragment(), ContractHome.View {

    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!
    private var totalDeProdutosDoUtilizador = 0

    override lateinit var presenter: ContractHome.Presenter

    private lateinit var adapterHome: AdapterHome
    private var currentCategories: List<PostCategory> = emptyList() // Para o dialog do FAB

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        // Inicialização do Room e DAOs
        val category = DependencyInjector.registerCategoryRepository(requireContext())
        val homeRepository = DependencyInjector.registerHomeRepository(requireContext())
        presenter = HomePresenter(this, homeRepository, category)

        binding?.let { b ->
            ViewCompat.setOnApplyWindowInsetsListener(b.fab) { v, insets ->
                val bars = insets.getInsets(WindowInsetsCompat.Type.navigationBars())
                val params = v.layoutParams as? ViewGroup.MarginLayoutParams
                params?.let {
                    // Define a margem inferior baseada na altura da barra de navegação + 16dp de respiro
                    it.bottomMargin = bars.bottom + 64
                    v.layoutParams = it
                }
                insets
            }
        }
        _binding?.let { b ->
            b.layoutEmptyStateAdd.setOnClickListener {
                // Simula o clique do botão FAB principal para abrir o diálogo de categorias
                b.fab.performClick()
            }
        }
        setupRecyclerView()
        // ✅ ATUALIZAÇÃO: Escuta dinamicamente a contagem total de produtos no banco de dados local
        val database = BpDatabase.getDatabase(requireContext())
        database.productDao().getTotalProductsCountLiveData().observe(viewLifecycleOwner) { count ->
            totalDeProdutosDoUtilizador = count ?: 0
        }
        setupFab()
        observePresenterFlows()
        // ── ✅ CORREÇÃO: Lê os argumentos para aplicar filtro de categoria se houver ──
        val categoryId = arguments?.getString("categoria_id").orEmpty()
        val categoryNome = arguments?.getString("categoria_nome").orEmpty()

        if (categoryId.isNotEmpty()) {
            // Atualiza o título da Toolbar para mostrar que a categoria está ativa
            (activity as? AppCompatActivity)?.supportActionBar?.title = "${
                requireContext().getString(R.string.category)
            }: $categoryNome"
            presenter.fetchProductsByCategory(categoryId)
        } else {
            // Comportamento normal caso venha do menu lateral direto
            presenter.fetchProducts()
        }

        presenter.fetchCategories()
        // Se não há categoria, fetchProducts() já carregou tudo normalmente
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onDestroy() {
        if (::presenter.isInitialized) presenter.onDestroy()
        super.onDestroy()
    }

    private fun setupRecyclerView() {
        val diasAmarelo = NotificationPrefs.getDays(requireContext())

        adapterHome = AdapterHome(
            requireContext(),
            diasAmarelo,
            object : OnItemClickListener {

                override fun onItemClick(position: Int, data: List<PostProduct>) {
                    val produto = data.getOrNull(position) ?: return
                    val bundle = Bundle().apply {
                        putString("uuid", produto.uuid)
                        putParcelable("product_bundle", produto)
                    }
                    findNavController().navigate(
                        R.id.action_nav_home_to_nav_edit_fragment,
                        bundle
                    )
                }

                override fun onImageClick(uri: String) {
                    val bundle = Bundle().apply { putString("imageUri", uri) }
                    findNavController().navigate(
                        R.id.action_nav_home_to_nav_show_fragment,
                        bundle
                    )
                }
                override fun onHeaderDeleteClick(productsToDelete: List<PostProduct>, groupTitle: String) {
                    val total = productsToDelete.size
                    val mensagem = if (total == 1) {
                        getString(R.string.delete_confirm_single, productsToDelete.first().name)
                    } else {
                        getString(R.string.delete_confirm_multiple, total, groupTitle)
                    }

                    DialogUtil.showDeleteDialog(requireContext(), mensagem) {
                        CustomProgressBar.showLoadingDialog(requireContext())

                        // Executa a deleção sequencialmente para cada item do lote
                        productsToDelete.forEach { product ->
                            presenter.deleteProduct(product)
                        }

                        // Temporizador simples para ocultar o progresso
                        android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
                            CustomProgressBar.hideLoadingDialog()
                            // O fluxo reativo de dados atualizará a UI automaticamente
                        }, 1000)
                    }
                }
            }
        )

        binding.rcViewListHome.layoutManager = LinearLayoutManager(requireContext())
        binding.rcViewListHome.adapter = adapterHome
    }

    private fun setupFab() {
        binding.fab.setOnClickListener {
            val isPremium = SharedPreferencesManager.isPremium(requireContext())

            // ✅ ATUALIZAÇÃO: Bloqueia com base no total absoluto (ativos + lixeira) do banco de dados
            if (!isPremium && totalDeProdutosDoUtilizador >= 50) {
                mostrarDialogoLimitePremium()
                return@setOnClickListener
            }

            // Fluxo normal de abertura do diálogo de categorias
            CategoriaDialogUtil.mostrarDialogo(
                requireContext(),
                currentCategories,
                object : CategoriaDialogUtil.CategoriaCallback {

                    override fun onCategoriaSelecionada(categoriaId: String) {
                        val nome = currentCategories
                            .firstOrNull { it.firestoreId == categoriaId }?.name ?: ""
                        navegarParaScanner(categoriaId, nome)
                    }

                    override fun onAdicionarCategoria() {
                        AlertDialog.Builder(requireContext())
                            .setTitle(getString(R.string.dialog_title_no_category))
                            .setMessage(getString(R.string.dialog_msg_start_managing))
                            .setPositiveButton(getString(R.string.dialog_add_category_title)) { dialog, _ ->
                                dialog.dismiss()
                                findNavController().navigate(R.id.nav_category)
                            }
                            .setNegativeButton(getString(
                                R.string.dialog_button_not_now
                            )) { dialog, _ ->
                                dialog.dismiss()
                            }
                            .show()
                    }
                }
            )
        }
    }

    private fun observePresenterFlows() {
        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    (presenter as HomePresenter).products.collect { products ->
                        if (products != null) {
                            adapterHome.setDados(products)
                            showEmpty(products.isEmpty())
                        }
                    }
                }
                launch {
                    (presenter as HomePresenter).categories.collect { categories ->
                        currentCategories = categories
                    }
                }
            }
        }
    }

    private fun navegarParaScanner(categoriaId: String, categoriaNome: String) {
        val bundle = Bundle().apply {
            putString("categoria_id", categoriaId)
            putString("categoria_nome", categoriaNome)
        }
        findNavController().navigate(
            R.id.action_nav_home_to_nav_barcode_scan_fragment,
            bundle
        )
    }

    override fun showProgress(enable: Boolean) {
    }

    override fun showEmpty(isEmpty: Boolean) {
        val target = if (isEmpty) 1 else 0
        if (binding.viewFlipper.displayedChild != target) {
            binding.viewFlipper.displayedChild = target
        }
    }

    private fun mostrarDialogoLimitePremium() {
        AlertDialog.Builder(requireContext())
            .setTitle(getString(R.string.premium_limit_title))
            .setMessage(getString(R.string.premium_limit_message))
            .setPositiveButton(getString(R.string.premium_limit_positive)) { dialog, _ ->
                dialog.dismiss()
                findNavController().navigate(R.id.nav_payment)
            }
            .setNegativeButton(getString(R.string.premium_limit_negative)) { dialog, _ ->
                dialog.dismiss()
            }
            .show()
    }

    override fun onSuccess(name: String) {
        val b = _binding ?: return
        Snackbar.make(b.root,
            getString(R.string.product_deleted_success, name),
            Snackbar.LENGTH_SHORT).show()
    }

    override fun onError(message: String) {
        val b = _binding ?: return
        Snackbar.make(b.root, message, Snackbar.LENGTH_LONG).show()
    }

    override fun onProductDeleted(product: PostProduct) {
        onSuccess(getString(R.string.product_deleted_success, product.name))
    }

    override fun onProductRestored(product: PostProduct) {
        onSuccess(getString(R.string.product_deleted_restored, product.name))
    }
}